File: android/app/src/main/java/com/example/ubapp/realtime/StateMachine.kt
package com.example.ubapp.realtime

abstract class State {
    open fun enter() {}
    open fun exit() {}
    open fun update(dt: Float) {}
}

class StateMachine(initial: State) {
    var current: State = initial
        private set

    init {
        initial.enter()
    }

    fun transition(to: State) {
        if (to === current) return
        current.exit()
        current = to
        to.enter()
    }

    fun update(dt: Float) {
        current.update(dt)
    }
}

File: lib/realtime/state_machine.dart
abstract class GameState {
  void onEnter() {}
  void onExit() {}
  void update(double dt) {}
}

class StateMachine {
  StateMachine(this._current) {
    _current.onEnter();
  }

  GameState _current;
  GameState get current => _current;

  void transition(GameState next) {
    if (identical(next, _current)) return;
    _current.onExit();
    _current = next;
    next.onEnter();
  }

  void update(double dt) => _current.update(dt);
}
